### PR TITLE
BestPractices.md: Added guidance on how unused parameters in Get-TargetResource (fixes #206)

### LINKED ISSUE #206: How should we use and document unused non-mandatory parameter in the Get-method of a resource
I came across two different scenarios when unused non-mandatory parameters are used in the Get-method in my recent reviews.
What I mean with 'unused' is that they are no used for any validation or assignment in the Get-method (other than the returned hastable), but rather to be able to call Get-method using `Get-TargetResource @PSBoundParameter`.

In PR PowerShell/xSQLServer#147 the non-mandatory parameters are not used to determine outcome. And the paramater variable are later re-assigned to the value of the actual state, so the value provided by the user in the function call is never used in the returned hash table anyway.

Also, if allowing unused non-mandatory parameters in the Get-method and also using them in the returned hash table, do we not open up for potential giving false information to the user when calling the `Get-DscConfiguration`? We are returning the state the user wants for that parameter, not the actual state.

And in tests this would add more test code which gives even more complexity.

In PR PowerShell/xSQLServer#147 this could be the exception to the "rule". They are also added so the Get-method can be called using `Get-TargetResource @PSBoundParameter`.
But the two parameters in question (`RestartService` and `RestartTimeout`) does actually provide the user with relevant information which would be of interest for the user when calling `Get-DscConfiguration`.
In this PR @mbreakey3 asked to get comments on the unused parameters in the comment-based help. 
So there is a comment that says 

```
.PARAMETER RestartTimeout
*** Not used in this function ***
The length of time, in seconds, to wait for the service to restart. 
```

But that is not entirely true either since it is relevant information, and it is used in the returned hash table.

Anyway. So if we do allow unused non-mandatory parameters I think we need to add the approach that @mbreakey3 took above into the style guideline, or best practices guideline.

If I could choose I would like to _not_ allow unused parameters, but do allow the exceptions to the rule. :smile:

Just my thoughts. Happy to go either way.

### PR DESCRIPTION
**Pull request by johlju**
* Changes to bestpractices
  - Added guidance on how unused parameters in Get-TargetResource as per issue #206
  - Fixed lint errors
  - Added sub headings to table of contens

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/dscresources/229)

### CODE REVIEW

File: BestPractices.md [hunk 1]
```diff
@@ -416,153 +517,368 @@ function Get-Property
...
+
+    return @{
+        Name = $Name
+        Servername = $ServerName
+        Feeature = $feature
+    }
+}
+```
+
+### Get-TargetResource with unused mandatory parameters should include "Not used in Get-TargetResource" in the comment-based help parameter description
```
Comment (TravisEz13) [inline BestPractices.md]: Can you include a description of why it`s appropriate not to use a mandatory parameter in a get?

For example, in the example you provide Ensure describes a state of the resource which is mandatory, but should not be needed to retrieve the resource.

**Comment by johlju** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[BestPractices.md, line 715 at r1](https://reviewable.io:443/reviews/powershell/dscresources/229#-KWxWFfwakIdvqdonKLh-r1-715:-KZ6FVp5LI5H4RsRanJj:bb6un1j) ([raw file](https://github.com/powershell/dscresources/blob/890bba354c1e3a83d6ec2e48b511dc661c225848/BestPractices.md#L715)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

Can you include a description of why it`s appropriate not to use a mandatory parameter in a get?

For example, in the example you provide Ensure describes a state of the resource which is mandatory, but should not be needed to retrieve the resource.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229)*

**Comment by johlju** [comment]
@TravisEz13 Added more descriptive text.

@kwirkykat would you mind proof read?

**Comment by mbreakey3** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[BestPractices.md, line 721 at r2](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZ7R5gNRsbql55U4y6d:-KZ7R5gNRsbql55U4y6e:b-vyfj34) ([raw file](https://github.com/powershell/dscresources/blob/eeaae79befe263b2960d6c72a0e6075095d59879/BestPractices.md#L721)):*
> ```Markdown
> ```
> 
> ### Get-TargetResource with unused mandatory parameters should include "Not used in Get-TargetResource" in the comment-based help parameter description
> ```

This also comes up in Test-TargetResource for both mandatory and non-mandatory. So maybe make this rule more general: "Any unused parameters that must be included in a function definition should include 'Not used in <function_name>' in the help comment for that parameter above the function"

---

*[BestPractices.md, line 725 at r2](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZ7QqWGHPebPqQ11n9K:-KZ7QqWGHPebPqQ11n9L:b-llj722) ([raw file](https://github.com/powershell/dscresources/blob/eeaae79befe263b2960d6c72a0e6075095d59879/BestPractices.md#L725)):*
> ```Markdown
> The inclusion of a mandatory parameter that is never used could signal that there is a design flaw in the implementation of the Get-TargetResource function.
> The mandatory parameters that are used to call Get-TargetResource should help to retrieve the actual values of the properties for the resource.
> For example, if there is a parameter `Ensure` that is mandatory, that parameter will might not be used to retrive the actual values.
> ```

'that parameter will not be used...'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229)*

**Comment by johlju** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[BestPractices.md, line 721 at r2](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZ7R5gNRsbql55U4y6d:-KZ7evGrYGUahY0db2vG:bb6un1j) ([raw file](https://github.com/powershell/dscresources/blob/eeaae79befe263b2960d6c72a0e6075095d59879/BestPractices.md#L721)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

This also comes up in Test-TargetResource for both mandatory and non-mandatory. So maybe make this rule more general: "Any unused parameters that must be included in a function definition should include 'Not used in <function_name>' in the help comment for that parameter above the function"
</blockquote></details>

Done

---

*[BestPractices.md, line 725 at r2](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZ7QqWGHPebPqQ11n9K:-KZ7ewPlMpoXPR3AWsFJ:b-3lqan7) ([raw file](https://github.com/powershell/dscresources/blob/eeaae79befe263b2960d6c72a0e6075095d59879/BestPractices.md#L725)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

'that parameter will not be used...'
</blockquote></details>

Done (I think)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229)*

**Comment by mbreakey3** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[BestPractices.md, line 723 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIncvelXvLR7fZsMS:-KZOIncvelXvLR7fZsMT:b-q2ov00) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L723)):*
> ```Markdown
> ```
> 
> ### Any unused parameters that must be included in a function definition should include 'Not used in `function_name`' in the help comment for that parameter in the comment-based help
> ```

Not used in <function_name>.

---

*[BestPractices.md, line 724 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIuwqmDYerazKBd_r:-KZOIuwqmDYerazKBd_s:b-p5y719) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L724)):*
> ```Markdown
> 
> ### Any unused parameters that must be included in a function definition should include 'Not used in `function_name`' in the help comment for that parameter in the comment-based help
> The inclusion of a mandatory parameter in the `Get-TargetResource` function that is never used could signal that there is a design flaw in the implementation
> ```

can you switch the quotes to 'regular single quotes' instead of `backticks` ?

---

*[BestPractices.md, line 727 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOHTo-b-hyw-KQYWTI:-KZOHTo-b-hyw-KQYWTJ:b-9fur5g) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L727)):*
> ```Markdown
> of the function.
> The mandatory parameters that are used to call `Get-TargetResource` should help to retrieve the actual values of the properties for the resource.
> For example, if there is a parameter `Ensure` that is mandatory, that parameter will might not be used to retrive the actual values.
> ```

'that parameter will not be used to retrieve the actual values.'

---

*[BestPractices.md, line 733 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOHtYQFk14j_ZBVFJa:-KZOHtYQFk14j_ZBVFJb:b-2zxfaw) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L733)):*
> ```Markdown
> 
> The inclusion of a mandatory or a non-mandatory parameter in the `Test-TargetResource` function that is not used will be more common since it is required that both
> the `Set-targetResource` and the `Test-TargetResource` must have the same parameters. And there will be times when all of the paramters in the `Test-TargetResource`
> ```

'The inclusion of a mandatory or a non-mandatory parameter in the `Test-TargetResource` function that is not used is more common since it is required that both
the 'Set-TargetResource' and the 'Test-TargetResource' have the same parameters. Thus, there will be times when not all of the parameters in the 'Test-TargetResource'
function will be used in the function.

---

*[BestPractices.md, line 736 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIR8R48gwn2BVr4Yj:-KZOIR8R48gwn2BVr4Yk:b-udhrsv) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L736)):*
> ```Markdown
> function will not be used for testing the state of the resource.
> 
> If there is a need design-wise to include a mandatory parameter that will not be used, then the comment-based help for that parameter should contain the description 'Not used in Get-TargetResource'.
> ```

'Not used in <function_name>

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229)*

**Comment by johlju** [comment]
@mbreakey3 sorry it took so long to fix those comments :)

---

Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[BestPractices.md, line 723 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIncvelXvLR7fZsMS:-K_ieRKC0vJpap8a0eun:b-896fix) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L723)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

Not used in <function_name>.
</blockquote></details>

Done.

---

*[BestPractices.md, line 724 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIuwqmDYerazKBd_r:-K_ieSOVOg7OMmuDPVGa:b-896fix) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L724)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

can you switch the quotes to 'regular single quotes' instead of `backticks` ?
</blockquote></details>

Done.

---

*[BestPractices.md, line 727 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOHTo-b-hyw-KQYWTI:-K_ieYYF5CyaY7dv86A-:b-896fix) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L727)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

'that parameter will not be used to retrieve the actual values.'
</blockquote></details>

Done.

---

*[BestPractices.md, line 733 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOHtYQFk14j_ZBVFJa:-K_iehgCp9WbJays5p_i:b-896fix) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L733)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

'The inclusion of a mandatory or a non-mandatory parameter in the `Test-TargetResource` function that is not used is more common since it is required that both
the 'Set-TargetResource' and the 'Test-TargetResource' have the same parameters. Thus, there will be times when not all of the parameters in the 'Test-TargetResource'
function will be used in the function.

</blockquote></details>

Done.

---

*[BestPractices.md, line 736 at r3](https://reviewable.io:443/reviews/powershell/dscresources/229#-KZOIR8R48gwn2BVr4Yj:-K_iesoXTH5D5p_yx2sz:b-896fix) ([raw file](https://github.com/powershell/dscresources/blob/37396582eed136ae2b57b1140ef63e683e44ed40/BestPractices.md#L736)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary><blockquote>

'Not used in <function_name>
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229#-:-K_ieDiMFhFQJ6t6aAl7:b78rhb1)*

**Comment by kwirkykat** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[BestPractices.md, line 8 at r4](https://reviewable.io:443/reviews/powershell/dscresources/229#-Kat7Bld5jzXVmjQj8fC:-Kat7Bld5jzXVmjQj8fD:b-gcj10c) ([raw file](https://github.com/powershell/dscresources/blob/549c1b112b139f4ab95e184fb8fdee0f99a643b2/BestPractices.md#L8)):*
> ```Markdown
> 
> - [General](#general)
>   - [Avoid Using Hardcoded Computer Name](#avoid-using-hardcoded-computer-name)
> ```

YES! Thanks for adding the table of contents :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresources/229#-:-Kat7i46m-0kFG33d27G:bnfp4nl)*

**Comment by kwirkykat** [comment]
Change requested by @TravisEz13 was made.

**Pull request merged**